### ISSUE TITLE
Failed to allocate low-address region during UsbDiscoverer.getZebraUsbPrinters()

### ISSUE THREAD

**Issue by tobiascapin** [icedevml/zxp-native-usb-adapter-64bit-hack#2]
Hi, first of all I want to thank you for this project. It is incredible that Zebra is not supporting new Java versions.

I'm experiencing some issue with your library, I could not understand why for some system it works and for some other it doesn't work.
My code uses `UsbDiscoverer.getZebraUsbPrinters()` from LinkOS SDK to disover usb printers, and it crashes.
I tried to add `DEBUG_MQALLOC=1○6 to env to get more details but I get just this output:

```
[MQALLOC] Loading ZebraNativeUsbAdapter_64.dll allocator compatibility hack by MQ
[MQALLOC] Build: 20231120R02
                       __________.__        __   __                __ ________  
   /\_/\               \____    /|__|__ ___/  |_|  | _______      / / \_____  \ 
   >^.^<.---.            /     / |  |  |  \   __\  |/ /\__  \    / /    _(__  < 
  _'-`-'     )\         /     /_ |  |  |  /|  | |    <  / __ \_  \ \   /       \
 (6--\ |--\ (`.`-.     /_______ \|__|____/ |__| |__|_ \(____  /   \_\ /______  /
     --'  --'  ``-'BP          \/                    \/     \/               \/ 
[MQALLOC] BUG! Failed to allocate low-address region.
```

Do you know what can be the reason of this bug or how can I get more informations?

I'm using Java 17.0.10+7 on x64 windows 10 machine.
The process is started inside a tomcat container with Local System account, I tried also Local Service or Network Service with no luck.

**Comment by tobiascapin** [comment]
I have a little followup, it looks like it is related to the windows service or apache commons daemon.

I tried to run the application directly from cmd using my user, and it works.
If I try to run it as windows service (tomcat10.exe, it uses apache commons daemon) it doesn't work even if the service is started with LocalService or with my personal user.

For some reason it seems `VirtualAlloc` method inside MQALLOC fails when started as windows service.

**Comment by tobiascapin** [comment]
Finally I could solve changing the startup method of tomcat daemon from jvm to exe. Seems that with a separated process MQALLOC is  working.

**Comment by kkko13** [comment]
Hello, I seem to have encountered the same problem as you. When I use IntelliJ Idea to run my printing project, I can perform printing tasks, but when I package it as a jar package and run it, there is an error in the “ZebraCardPrinterFactory. getInstance (connection) ” line. Do you have any idea about this？

**Comment by tobiascapin** [comment]
Do you have this issue from command line startup (out of IDE)?

**Comment by kkko13** [comment]
yeah, everything was normal when I ran the project in ide, but after I packaged the project into a jar package through maven, the printing failed. I was sure that the dll documents I needed and the official jar package were successfully packaged. What should I do with this situation?

**Comment by icedevml** [comment]
@tobiascapin The original issue is quite interesting. What the patch is doing is that it's trying to allocate a small chunk of virtual memory somewhere within that range of addresses:

```
#define MEM_REGION_SEARCH_START_ADDR  (0x05000000)
#define MEM_REGION_SEARCH_END_ADDR    (0x08000000)
```

Maybe there is something occupying those addresses already in Tomcat. I can potentially inspect that in spare time. What should I do to reproduce the issue? Just create a simple Tomcat Web Application and try to use ZXP SDK from within the app?

**Comment by icedevml** [comment]
> yeah, everything was normal when I ran the project in ide, but after I packaged the project into a jar package through maven, the printing failed. I was sure that the dll documents I needed and the official jar package were successfully packaged. What should I do with this situation?

@kkko13 I think the most probable case is that it's failing to load the required DLLs when you are running the packaged version. You would need to check how to package DLLs so that they are properly recognized within the classpath.

**Comment by icedevml** [comment]
@tobiascapin I'm unable to reproduce that one. Note that I'm not extremely familiar with Tomcat, but I've spun a test project using gretty gradle plugin with Tomcat 10 and implemented a simple `HttpServlet` for it. It works both from the IDE (through `./gradlew appRun`) and from standalone JAR-packaged setup.

In both cases the servlet is able to connect with printer and the memory is allocated without issues.

Nevertheless, the approach used in the patch is risky since there is absolutely no guarantee that some arbitrary range of virtual addresses will be free to allocate something on them. I have an idea how that patch could be reimplemented to be less hacky around that. Will implement that whenever there will be some spare time.

**Comment by icedevml** [comment]
Actually, I've got quite curious about having that implementation done, so here is some pre-release version with the improved allocation algorithm. Maybe that one will work correctly on that Tomcat setup. This one works fully correctly on my end with a ZXP printer (however I can't reproduce that specific issue that you were hitting with memory allocation).

https://github.com/icedevml/zxp-native-usb-adapter-64bit-hack/releases/tag/build-20250817R04

**Comment by icedevml** [comment]
I've checked with Tomcat 10 running as local service, it seems to work just fine now.
